File: 1482 Minimum Number of Days to Make m Bouquets/1482minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int ans = INT_MAX;

    int check(vector<int> arr,int mid,int k){
        int cnt=0;
        int cur=0;
        for(int i=0;i<arr.size();i++){
            if(arr[i]<=mid){
                cur+=1;
                if(cur==k){
                    cnt++;
                    cur=0;
                }
            }
            else{
                cur = 0;
            }
        }
        return cnt;
    }

    int minDays(vector<int>& bloomDay, int m, int k) {
        int s = *min_element(bloomDay.begin(),bloomDay.end());
        int e = *max_element(bloomDay.begin(),bloomDay.end());
        while(s<=e){
            int mid = s+(e-s)/2;
            int a1 = check(bloomDay,mid,k);
            if(a1<m){
                s = mid+1;
            }
            else if(a1>=m){
                ans = min(ans,mid);
                e = mid-1;
            }
        }
        if(ans == INT_MAX)
            return -1;
        return ans;
    }
};

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.py
class Solution:
    def minDays(self, bloomDay: list[int], m: int, k: int) -> int:
        
        if m*k > len(bloomDay):
            return -1
        
        def possible(val):
            bouquet = 0
            adjacency = 0

            for day in bloomDay:
                if day <= val:
                    adjacency += 1

                    if adjacency == k:
                        bouquet += 1
                        adjacency = 0

                        if bouquet >= m:
                            return True
                else:
                    adjacency = 0

            return False


        right = max(bloomDay)
        left = min(bloomDay)

        while left < right:
            mid = left + (right - left )//2

            if possible(mid):
                right = mid
            else:
                left = mid + 1
        
        return left
        

